drivers/ps2/ps2_npcx_controller.c

```diff
@@ -0,0 +1,354 @@
+/*
+ * Copyright (c) 2021 Nuvoton Technology Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nuvoton_npcx_ps2_ctrl
+
+/**
+ * @file
+ * @brief Nuvoton NPCX PS/2 module (controller) driver
+ *
+ * This file contains the driver of PS/2 module (controller) which provides a
+ * hardware accelerator mechanism to handle both incoming and outgoing data.
+ * The hardware accelerator mechanism is shared by four PS/2 channels.
+ */
+
+#include <drivers/clock_control.h>
+#include <drivers/ps2.h>
+#include <dt-bindings/clock/npcx_clock.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(ps2_npcx_ctrl, CONFIG_PS2_LOG_LEVEL);
+
+#define NPCX_PS2_CH_COUNT 4
+
+/*
+ * Set WDAT3-0 and clear CLK3-0 in the PSOSIG register to
+ * reset the shift mechanism.
+ */
+#define NPCX_PS2_SHIFT_MECH_RESET (uint8_t) ~NPCX_PSOSIG_CLK_MASK_ALL
+
+/* in 50us units */
+#define PS2_RETRY_COUNT 10000
+
+#define PS2_TRANSACTION_TIMEOUT K_MSEC(20)
+
+/* Device config */
+struct ps2_npcx_ctrl_config {
+	uintptr_t base;
+	struct npcx_clk_cfg clk_cfg;
+};
+
+/* Driver data */
+struct ps2_npcx_ctrl_data {
+	/*
+	 * Bit mask to record the enabled PS/2 channels.
+	 * Only bit[7] and bit[5:3] are used
+	 * (i.e. the bit position of CLK3-0 in the PS2_PSOSIG register)
+	 */
+	uint8_t channel_enabled_mask;
+	struct k_sem lock;
+	struct k_sem tx_sync_sem;
+	ps2_callback_t callback_isr[NPCX_PS2_CH_COUNT];
+};
+
+/* Driver convenience defines */
+#define DRV_CONFIG(dev) ((const struct ps2_npcx_ctrl_config *)(dev)->config)
+
+#define DRV_DATA(dev) ((struct ps2_npcx_ctrl_data *)(dev)->data)
+
+#define HAL_PS2_INSTANCE(dev) (struct ps2_reg *)(DRV_CONFIG(dev)->base)
+
+int ps2_npcx_ctrl_configure(const struct device *dev, uint8_t channel_id,
+			    ps2_callback_t callback_isr)
+{
+	struct ps2_npcx_ctrl_data *const data = DRV_DATA(dev);
+
+	k_sem_take(&data->lock, K_FOREVER);
+
+	if (channel_id >= NPCX_PS2_CH_COUNT) {
+		LOG_ERR("unexpected channel ID: %d", channel_id);
+		return -EINVAL;
+	}
+
+	if (callback_isr == NULL) {
+		return -EINVAL;
+	}
+
+	data->callback_isr[channel_id] = callback_isr;
+	k_sem_give(&data->lock);
+
+	return 0;
+}
+
+int ps2_npcx_ctrl_enable_interface(const struct device *dev, uint8_t channel_id,
+				   bool enable)
+{
+	struct ps2_npcx_ctrl_data *const data = DRV_DATA(dev);
+	struct ps2_reg *const inst = HAL_PS2_INSTANCE(dev);
+
+	k_sem_take(&data->lock, K_FOREVER);
+	/*
+	 * Disable the interrupt during changing the enabled channel mask to
+	 * prevent from preemption.
+	 */
+	irq_disable(DT_INST_IRQN(0));
+
+	if (channel_id >= NPCX_PS2_CH_COUNT) {
+		LOG_ERR("unexpected channel ID: %d", channel_id);
+		return -EINVAL;
+	}
+
+	if (enable) {
+		data->channel_enabled_mask |= BIT(NPCX_PSOSIG_CLK(channel_id));
```

Comment:
You use BIT(NPCX_PSOSIG_CLK(channel_id)) a lot

How about adding a function that returns the mask for a particular channel's clock?

Reply:
Done by adding a function to return the bit-mask.